Search from navbar searchform

clukhei opened the pull request:
1) Search by name on navbar 
- on enter , fetch API, direct to page /search/:name 
- data is passed down to SearchResults component

2) SearchResults component
Header: 
- no. of donors found 
- reiterate what the user searched for 

Table;
- donors and details 
- hover and click on rows will direct user to donor details page 
- pagination appears when results >15 
- handle null and falsy data 

future works: 
- sort table

clukhei committed:
set up searchresults page, direct to searchresults page after fetch API, pass data to searchresults page using history.push state prop and useLocation

clukhei committed:
pass data to searchresult table, searched name to be reflected on searchresults page using useParams()

clukhei committed:
add pagination to table, handleFalsy and handleBoolean data, onclick table rows direct to donor detail page

SheldonCheng commented:
@clukhei 👍 we have search now!

Something about your implementation: You are doing the fetch in the navbar and passing the result into history state.
This is kind of an unusual pattern though. If i try to access [http://localhost:3000/search/teo](http://localhost:3000/search/teo) directly by entering it into the URL bar, it does not work.

I would move the fetch into the SearchResults page instead, inside a `useEffect` hook. You already have the search string with `searched.name`. With this approach you can also add a spinner/loader while the data is loading.

clukhei committed:
move fetch to SearchResults,create spinner lib, spinner implemented in searchresults, dashboard, upload result and donor details page

clukhei committed:
fix CI build failure due to pagination path

clukhei commented:
@SheldonCheng 
Changes: 

-Moved fetch to SearchResults component  
- created bootstrap spinner.js lib 
-implemented spinner for SearchResults, Dashboard, upload success page, DonorDetails page, 
- change calendar date format to dd-mm-yyyy (dashboard calendar and donor list filters calendar)
- minor UI edits 
-move pagination.js and dateStringGenerator.js to lib folder 


To do (after this PR?): 
1) resolve FE issue when empty field is submitted to search API : 
- now showing blank page when empty field is searched...

SheldonCheng commented on src/components/DonorDetails/index.css:
```diff
@@ -38,6 +38,7 @@
   width: inherit;
   justify-content: center;
   align-content: 'space-around';
```
typo here, shouldn't have quotes `' '` with `'space-around'`

SheldonCheng commented on src/components/DonorDetails/index.js:
```diff
@@ -48,10 +49,10 @@ function DonorDetails(props) {
       </div>
     )
-  } else return null
+  } else return <Spin />
```
This can look nicer with the `? ... : ...` approach that you are already using in `Dashboard/index.js`

SheldonCheng commented on src/components/SearchResults/index.js:
```diff
@@ -0,0 +1,61 @@
+            </Header.Content>
+          </Header.Top>
+        </Header>
+        <ResultsTable data={searchResults} />
```
Can add a condition that checks `searchResults.length` is greater than `0` before it shows `ResultsTable`. Right now if there are no results, it's an empty table with the header still visible.

![image](https://user-images.githubusercontent.com/3888250/77565551-58938480-6eff-11ea-85a3-4dc3d0c5c9ba.png)

SheldonCheng commented on src/components/UpdateDB/Success.js:
```diff
@@ -23,119 +24,124 @@ const SuccessUpload = props => {
-    </div>
-  )
+    )
+  } else return <Spin />
```
same thing with `? .. : ..`

SheldonCheng reviewed (approved):
Nicely done, just some minor code formatting tweaks and hiding a table when it is empty.

Empty search term: I understand you are going to work on this next. Some hints:
- it should at least look the same as "no results found" 
- If you want can go one step further: if search term is empty, hint to the user to enter a search term.

clukhei committed:
minor code formatting, add spinner for DonorList

clukhei committed:
disable search when empty field is submitted

clukhei committed:
Remove preventDefault

can be removed now because api is no longer fetched in navbar component

allow form state to be reset every time a form is submitted

 this will help to load spinner when trying to searh from searchResults page

clukhei committed:
add back preventDefault() only when empty search is entered to prevent page from reloading when there is nothing searched
